### PR TITLE
Admin Page: Show better error messages when we can't parse JSON from API responses

### PR DESCRIPTION
**Pull request by oskosk**
Fixes #

#### Changes proposed in this Pull Request:

* Updates message to be shown when JSON cannot be parsed from a REST API response.

#### Testing instructions:

* Checkout this branch and build. Or test it using Jetpack Beta for this PR.
* Play with manually generated errors by using this snippet in line `331` of `_inc/lib/core-api/class.jetpack-core-api-module-endpoints.php`
  ```
            echo '/';
            //http_response_code(404);
            //header('Location: /');
            //header('Location: /nonexistantpath');
            exit;
  ```
* By uncommenting the first one, you'll get a JsonParseError. (Resembling something broken in the user's WP installation but unrelated to Jetpack
* By uncommenting the second one, you'll get a JsonApi404Error. (Resembling something weird in WP's config that doesn't let us reach the REST api.
* By uncommenting the third one, you'll get a redirection to the home. So the page will exist, but it's not where we need to end with a REST API query. 
* By uncommenting the 4th one, you'll get redirected to a 404. This resembles bad hosting config maybe coming from .htaccess or a reverse proxy. 

#### Proposed changelog entry for your changes:

* Introduces custom JS errors in `rest-api/index.js`
* Updates the updateSettings action to show `error.name` instead of `error.
* Adds better handling of errors thrown in fetch api promises

#### Screenshots.:

**When we can't parse the response from the API**

image

**When we can't parse the response from the API after detecting a redirect**

image

**When we get a 404 from calling an API endpoint**

image

**When we get a 404 from calling an API endpoint after detecting a redirect**

image

### CODE REVIEW

File: _inc/client/rest-api/index.js [hunk 1]
```diff
@@ -258,3 +258,7 @@ function checkStatus( response ) {
 		throw error;
 	} );
 }
+
+function catchJsonParseError( e ) {
+	throw new Error( `Couldn't understand Jetpack's REST API response (${ e.name })` );
```
Comment (zinigor) [inline _inc/client/rest-api/index.js]: Since this is going to face the user, let's rephrase please and use transatable strings here?

**Comment by oskosk** [comment]
@zinigor I'm not super sure if these strings are the way to go actually... The actual message is shown by an upper level of code with a translated `Updating settings failed` (the redux action that creates the notice). I've been playing last night with different messages and I'm not able to decide if any string we could show there would be useful per se for the user. 

I can only think that showing a message here about all of these new messages is that they're only useful if:
1. They help HEs debug issues
2. They help users find message in forums.

So, basically here's the flow of the strings

`small-api-client [request fails for any reason, throws JS error] -> redux action catches it (here the notice is created -> User -> HEs.`

current errors handled and introduced by this PR:

1. 404 after an API request gets redirected by the site/nginx/apache
1. Error parsing expected JSON when the response returns after being redirected by the site/nginx/apache
1. Error parsing expected JSON for any reason without being redirected (like, headers already sent message, or weird character left by the user in any php file processed by WordPress).

@jeherve what would you suggest that would make it better for all of us ?

**Comment by zinigor** [comment]
@oskosk yes, that's what I meant initially - we need to translate a general error message that will tell the user that something went wrong. Everything else will be the debugging output to be used in support requests.

**Comment by jeherve** [comment]
I like this! I wonder if we could add a link to contact support inside the notices.

> they're only useful if:
>
> They help HEs debug issues
> They help users find message in forums.

I would agree. It would be hard to come up with a default solution for those errors, but the errors will at least help HEs come up with the right next steps in helping the user.

**Comment by oskosk** [comment]
> I like this! I wonder if we could add a link to contact support inside the notices.

@jeherve where would you link to ?

**Comment by jeherve** [comment]
> EDIT: https://jetpack.com/contact-support?rel=support would be appropriate maybe

Yes, in this case they won't find anything helpful in the support docs, so the direct URL to the form (including `?rel=support`) would be more appropriate I think.

**Pull request merged**